static/app/components/profiling/flex.tsx

```diff
@@ -0,0 +1,84 @@
+import {CSSProperties} from 'react';
+import styled from '@emotion/styled';
+
+const px = (val: string | number | undefined) =>
+  typeof val === 'string' ? val : typeof val === 'number' ? val + 'px' : undefined;
+
+interface FlexProps {
+  align?: CSSProperties['alignItems'];
+  column?: boolean;
+  gap?: number | string;
+  h?: number | string;
+  justify?: CSSProperties['justifyContent'];
+  m?: number | string;
+  maxH?: number | string;
+  mb?: number | string;
+  minH?: number | string;
+  ml?: number | string;
+  mr?: number | string;
+  mt?: number | string;
+  mx?: number | string;
+  my?: number | string;
+  p?: number | string;
+  pb?: number | string;
+  pl?: number | string;
+  pr?: number | string;
+  pt?: number | string;
+  px?: number | string;
+  py?: number | string;
+  w?: number | string;
+  wrap?: CSSProperties['flexWrap'];
+}
+
+// TODO(@eliashussary): move to common folder / bring up in fe-tsc
+const FlexContainer = styled('div')<FlexProps>`
+  /* these can all come from a better base primitive */
+  display: flex;
+  height: ${p => px(p.h)};
+  width: ${p => px(p.w)};
+  min-height: ${p => px(p.minH)};
+  max-height: ${p => px(p.maxH)};
+  /* padding */
+  padding: ${p => px(p.p)};
+  padding-left: ${p => px(p.pl || p.px)};
+  padding-right: ${p => px(p.pr || p.px)};
+  padding-top: ${p => px(p.pt || p.py)};
+  padding-bottom: ${p => px(p.pb || p.py)};
+  /* margin */
+  margin: ${p => px(p.m)};
+  margin-left: ${p => px(p.ml || p.mx)};
+  margin-right: ${p => px(p.mr || p.mx)};
+  margin-top: ${p => px(p.mt || p.my)};
+  margin-bottom: ${p => px(p.mb || p.my)};
+  /* flex specific */
+  flex-direction: ${p => (p.column ? 'column' : 'row')};
+  justify-content: ${p => p.justify};
+  align-items: ${p => p.align};
+  gap: ${p => px(p.gap)};
+  flex-wrap: ${p => p.wrap};
+`;
+
+interface FlexItemProps {
+  basis?: CSSProperties['flexBasis'];
+  grow?: CSSProperties['flexGrow'];
+  shrink?: CSSProperties['flexShrink'];
+}
+
+const FlexItem = styled('div')<FlexItemProps>`
+  /* // TODO: determine sane defaults for these */
+  flex-grow: ${p => p.grow};
+  flex-shrink: ${p => p.shrink};
+  flex-basis: ${p => p.basis};
+  overflow: hidden;
+`;
+
+type FlexNamespace = typeof FlexContainer & {
+  Container: typeof FlexContainer;
+  Item: typeof FlexItem;
+};
+
+export const Flex = FlexContainer as FlexNamespace;
+Object.assign(FlexContainer, {
+  Container: FlexContainer,
+  Item: FlexItem,
+});
```

Comment:
Are there any benefits here between this and exporting an obj?
```js
export {
  Container: FlexContainer,
  Item: FlexItem
}
```

Reply:
An object isn't callable, the way I've done it is assigning Container and Item to Flex, while making Flex a reference to Container... imo it makes for cleaner usage and imports. 

By doing it this way we can encourage some namespaced components so that when you read it you're not reading it as an arbitrary 'Container' its 'Flex.Container' or simply 'Flex'. 

In addition to that it's related components you'll use to compose components are attached to its parent... 'Flex.Item'

It's a pattern I've successfully used in the past.

If I had to do it another way I'd simply just do 'FlexContainer' and 'FlexItem' but I'd only really reach for that if I had some serious tree shaking concerns. It's negligible here.

Reply:
The key difference is that in my instance i can simply call `<Flex />` because its callable, resulting in syntax that can be as follows;

```tsx
<Flex>
  <Flex.Item></Flex.Item>
</Flex>

// same as

<Flex.Container>
   <Flex.Item></Flex.Item>
</Flex.Container>
```


fwiw; i don't expect anyone really to explicitly use `Flex.Container` the pattern generally works better with components that are composed of a variety of related children;

```tsx
// in this case <Card/> has the same effect as <Card.Container/>
const FooCard = () => {
  return (
    <Card> 
      <Card.Header>
        <Card.Title>Foo</Card.Title>
      </Card.Header>
      <Card.Image />
      <Card.Content>Some Body</Card.Content>
      <Card.Footer>Bar</Card.Footer>
    </Card>
  );
};

```




Re: type cast... taking another glance over this, typecsript can properly infer the type if i simply assign `flex` the return of `Object.assign` - should've done this to begin with
```tsx
export const Flex = Object.assign(FlexContainer, {
  ...FlexContainer,
  Container: FlexContainer,
  Item: FlexItem,
});
```
![image](https://user-images.githubusercontent.com/7349258/215803607-72ef5211-e5ba-4ec8-9403-dff75a58286b.png)